### PR TITLE
Enable target imputation for sklearn multitask models

### PR DESCRIPTION
**Pull request by hesther**
Missing target values in sklearn multitask models can be imputed by specifying `--impute_mode <model/linear/median/mean/frequent>`. The option `model` trains single task sklearn models on each task to predict missing values and is computationally expensive. The option `linear` trains a stochastic gradient linear model on each target to compute missing targets. Both `model` and `linear` are applicable to regression and classification task. For regression tasks, the options `median` and `mean` furthermore compute the median and mean of the training data. For classification tasks, `frequent` computes the most frequent value for each task. For all options, models are fitted to non-missing training targets and predict missing training targets. The test set is not affected by imputing.

To test, you can for example use `tests/data/classification.csv` which has lots of missing targets, e.g. without imputation, single-task models:

```
python sklearn_train.py --data_path tests/data/classification.csv --dataset_type classification --model_type random_forest --single_task
```
which yields an AUC of 0.704. With imputation

```
python sklearn_train.py --data_path tests/data/classification.csv --dataset_type classification --model_type random_forest --impute_mode linear
```

an AUC of 0.735 is achieved for linear models (and 0.721 with `model`, 0.722 with `frequent`), which is slightly better since the model can make use of information that is shared between targets.

### CODE REVIEW

**Comment by hesther** [comment]
I tested imputation with all options against a few regression and classification tasks. I did not implement multiclass, since sklearn models cannot do multiclass anyways (and therefor raise a ValueError before model training anyways).

**Comment by hesther** [comment]
fixed the unused variable (from logging)

**Comment by cjmcgill** [comment]
Same as my comment in #211 I'm running into errors running the test cases with the dummy classification test set.

**Comment by cjmcgill** [comment]
> which yields an AUC of 0.704. With imputation an AUC of 0.735 is achieved for linear models (and 0.721 with model, 0.722 with frequent),

I'm afraid that it looks like sklearn doesn't have calculation seeds set up like pytorch, so that it's not perfectlly repeatable.

File: chemprop/args.py [hunk 1]
```diff
@@ -749,6 +749,8 @@ class SklearnTrainArgs(TrainArgs):
     """Number of bits in morgan fingerprint."""
     num_trees: int = 500
     """Number of random forest trees."""
+    impute_mode: Literal['model', 'median', 'mean', 'linear','frequent'] = None
```
Comment (cjmcgill) [inline chemprop/args.py:752]: The option `model` is not an especially clear name for what is being done. We could replace it with `single_task` or `single_task_prediction` for more clarity.

Comment (hesther) [reply chemprop/args.py:752]: Agree, will change it!

**Comment by hesther** [comment]
> Same as my comment in #211 I'm running into errors running the test cases with the dummy classification test set.
> 
> Edit: oh these are just warnings, it is in fact completing. The problem is that the random split is removing all the positive examples of the second task from the training set.

Yeah, looks like these are warnings from empty slices depending on the data splits. I don't think we need to do something about it, with a real dataset (not the dummy test one), it should be fine (I tested on e.g. QM9 and QM8 subsets).

**Comment by cjmcgill** [comment]
> > which yields an AUC of 0.704. With imputation an AUC of 0.735 is achieved for linear models (and 0.721 with model, 0.722 with frequent),
> 
> I'm afraid that it looks like sklearn doesn't have calculation seeds set up like pytorch, so that it's not perfectlly repeatable.

Oh! This is actually fixable. Line 323 in `sklearn_train.py` is missing the argument `random_state=args.seed`, which fixes this when I add it in. This was only a classification problem apparently.

**Comment by hesther** [comment]
> > > which yields an AUC of 0.704. With imputation an AUC of 0.735 is achieved for linear models (and 0.721 with model, 0.722 with frequent),
> > 
> > 
> > I'm afraid that it looks like sklearn doesn't have calculation seeds set up like pytorch, so that it's not perfectlly repeatable.
> 
> Oh! This is actually fixable. Line 323 in `sklearn_train.py` is missing the argument `random_state=args.seed`, which fixes this when I add it in. This was only a classification problem apparently.

Nice, we should add the seed to the current PR

**Comment by hesther** [comment]
Changed "model" to "single_task", and added the correct seed to the classifier (why was it only missing there?)

File: chemprop/sklearn_train.py [hunk 2]
```diff
@@ -136,6 +208,17 @@ def multi_task_sklearn(model: Union[RandomForestRegressor, RandomForestClassifie
...
+
+    if args.impute_mode:
+        train_targets = impute_sklearn(model=model,
+                                       train_data=train_data,
+                                       args=args,
+                                       logger=logger)
+    elif any(None in sublist for sublist in train_targets):
+        raise ValueError("Missing target values not tolerated for multi-task sklearn models." 
+                         "Use either --single_task to train multiple single-task models or impute"
+                         " targets via --impute_mode  <model/linear/median/mean/frequent>.")
```
Comment (cjmcgill) [inline chemprop/sklearn_train.py:220]: We could have it revert to single_task automatically if we wanted to. But now that we have impute_targets as a better option I don't think we'd ever want someone too. So this meaningful error with the options spelled out is great.

File: chemprop/sklearn_train.py [hunk 3]
```diff
@@ -55,6 +57,76 @@ def predict(model: Union[RandomForestRegressor, RandomForestClassifier, SVR, SVC
...
+                                              for features, targets in zip(train_data.features(), train_data.targets())
+                                              if targets[task_num] is not None])
+            if args.impute_mode == 'single_task':
+                model.fit(train_features, train_targets)
+                impute_train_preds = predict(
+                    model=model,
+                    model_type=args.model_type,
+                    dataset_type=args.dataset_type,
+                    features=impute_train_features
+                    )
```
Comment (cjmcgill) [inline chemprop/sklearn_train.py:101]: Unindent this ')' one tab

**Review by cjmcgill — APPROVED** [review]
Add in the fix to logging that I mentioned in #211 and you'll be good to go. Going ahead and approving. Once you make that change you can go ahead and merge.

**Comment by hesther** [comment]
Included all changes, waiting for the tests to finish and will merge it then. Great catch with the pformat logging issues!
